### ISSUE TITLE
Init startup 4.7 is slow (Favorites - fixed, GPX - investigating, Native files - reproducing)

### ISSUE THREAD

**Issue by sonora** [osmandapp/OsmAnd#19373]
On my usual setup an apk built from our master code on 2024-03-15 takes 7 sec between tapping the app icon on the phone's home screen and the map becoming fully visible. (App banner shows for 4 sec, then a gray map screen with initially only the relief layer shows for another 3 sec).

An analog build from our code on 2024-03-22 takes about 40 sec, with the banner dissappearing after about 6 sec and the rest spent on the map being shown empty with initially only the relief layer visible.

The effect is completely reproducible when switching back and forth between the builds.

Setup:  V1 engine, classical relief layer on with some favs and ~50 tracks visible.

**Comment by sonora** [comment]
PS: With older builds, I lately had the effect (reported elsewhere already) that the map screen became unresponsive shortly after each fresh startup, for about 25 sec or so, sometimes Android 11 reported an stalled app during this offering to wait or close and sometimes not.
But in all cases this fell into a period when the map was already fully displayed.

It seems that with new builds now this falls into the period where the map is still shown gray with only the relief layer visible, so the startup timing has somehow changed.

But in any case it seems the root cause for the app always stalling must be researched and eliminated. So I guess the core issie is **"App becomes unresponsive for about 25 sec shortly after each fresh startup"**.)

I usually have quite a siezeable number of maps (several dozens) active, not sure if that is related.

**Comment by sonora** [comment]
PS: I managed to get rid of the intermediate 'App not responsive' situations by temporarily (via renaming in the file system) using an **empty** osmand data folder. After renaming back, that nuisance was gone, so something seems to have now reset ok (in an internal app cache or similar?) This now shortened the startup time of the new OsmAnd builds to 'only' about 25 sec, with no more freezing, but still about 3 times longer than if going back to my March 15 build).

**Comment by vshcherb** [comment]
>> On my usual setup an apk built from our master code on 2024-03-15 takes 7 sec

Well since I switched to release build I can say all my issues with slowness are gone, app starts 3 seconds instead of 15 seconds. Search is instant (not everywhere probably), so I definitely see an improvement. 

We plan to share a test harness with GPX to measure apple to apples. So you can you do measurements using TET gpx, https://transeurotrail.org/poland/#5.89/52.967/18.004. 

Though still I suspect non-release build

**Comment by sonora** [comment]
I will keep an eye on this. From what I see, I have already exclusively been using what you call 'release' builds for all my testing in recent years, long before we recently switched our nightlies to use release code. Because many years ago already I had noticed the pronounced slugginess of builds using debug libs and debug compilation. My builds had also always been several MB smaller than our comparable nightlies, and only after the recent nightly switch  they are now both within a few hundred bytes again, so I suspect they are equivalent.

In any case that would not explain the reproducible deterioration I see between my builds of March 15 and March 21, as both are compiled alike. Will see if the slowness of the new builds persists.

**Comment by sonora** [comment]
@vshcherb Here's the log files for the startups, for comparison:
[logcat_2024-03-15.log.txt](https://github.com/osmandapp/OsmAnd/files/14754493/logcat_2024-03-15.log.txt)
[logcat_2024-03-25.log.txt](https://github.com/osmandapp/OsmAnd/files/14754494/logcat_2024-03-25.log.txt)

Looks like the new builds spend a lot of time on
`System.out: Initialized FAVORITES_INITIALIZED in 15909 ms
`
and
`net.osmand: MapRenderRepositories Native search: 15529 ms  {Loader map objects (synchronizer)}
`

vs before
`System.out: Initialized NATIVE_INITIALIZED in 2999 ms
`
and
`net.osmand: MapRenderRepositories Native search: 146 ms  {Loader map objects (synchronizer)}
`
??

**Comment by vshcherb** [comment]
Can confirm that favorites initialization is pretty slow:
```
03-26 14:05:18.495 19487 19551 I System.out: Initialized INIT_RENDERERS in 107 ms
03-26 14:05:19.146 19487 19487 I System.out: Initialized ROUTING_CONFIG_INITIALIZED in 651 ms
03-26 14:05:19.456 19487 19551 I System.out: Initialized NATIVE_OPEN_GL_INITIALIZED in 309 ms
03-26 14:05:19.487 19487 19551 I System.out: Initialized POI_TYPES_INITIALIZED in 32 ms
03-26 14:05:19.503 19487 19551 I System.out: Initialized ASSETS_COPIED in 16 ms
03-26 14:05:19.617 19487 19551 I System.out: Initialized MAPS_INITIALIZED in 114 ms
03-26 14:05:20.110 19487 19551 I System.out: Initialized NATIVE_INITIALIZED in 493 ms
03-26 14:05:20.640 19487 19551 I System.out: Initialized FAVORITES_INITIALIZED in 530 ms
03-26 14:05:20.659 19487 19551 I System.out: Initialized POI_FILTERS_INITIALIZED in 19 ms
03-26 14:05:21.058 19487 19551 I System.out: Initialized INDEX_REGION_BOUNDARIES in 399 ms
03-26 14:05:21.182 19487 19551 I System.out: Initialized LOAD_GPX_TRACKS in 124 ms
03-26 14:05:21.186 19487 19551 I System.out: Initialized SAVE_GPX_TRACKS in 4 ms
03-26 14:05:21.186 19487 19551 I System.out: Initialized RESTORE_BACKUPS in 0 ms
```
vs
```
03-26 08:34:40.214 24871 24871 I System.out: Time to start application 1700 ms. Should be less < 800 ms
03-26 08:34:40.286 24871 24871 I System.out: Time to init plugins 72 ms. Should be less < 800 ms
03-26 08:34:40.659 24871 26372 I System.out: Initialized INIT_RENDERERS in 444 ms
03-26 08:34:40.844 24871 26372 I System.out: Initialized POI_TYPES_INITIALIZED in 185 ms
03-26 08:34:40.922 24871 26372 I System.out: Initialized ASSETS_COPIED in 78 ms
03-26 08:34:41.814 24871 26372 I System.out: Initialized MAPS_INITIALIZED in 892 ms
03-26 08:34:42.254 24871 24871 I System.out: Initialized ROUTING_CONFIG_INITIALIZED in 440 ms
03-26 08:34:45.253 24871 26372 I System.out: Initialized NATIVE_INITIALIZED in 2999 ms
03-26 08:34:52.203 24871 26372 I System.out: Initialized FAVORITES_INITIALIZED in 6950 ms
03-26 08:34:52.258 24871 26372 I System.out: Initialized POI_FILTERS_INITIALIZED in 55 ms
03-26 08:34:54.243 24871 26372 I System.out: Initialized INDEX_REGION_BOUNDARIES in 1985 ms
03-26 08:34:54.243 24871 26372 I System.out: Initialized LOAD_GPX_TRACKS in 0 ms
03-26 08:34:54.248 24871 26372 I System.out: Initialized SAVE_GPX_TRACKS in 5 ms
03-26 08:34:54.248 24871 26372 I System.out: Initialized RESTORE_BACKUPS in 0 ms
```

**Comment by vshcherb** [comment]
To do: initialization of 10-50KB of favorites should be < 100 ms.

**Comment by sonora** [comment]
Yes, but it seems not just favorites, it's these two effects present in my March 25 log file, each adding about 16 sec, and much faster in the March 15 version:

`03-26 08:31:44.652 20221 24301 I System.out: Initialized FAVORITES_INITIALIZED in 15909 ms`
`03-26 08:31:44.766 20221 24297 I net.osmand: MapRenderRepositories Native search: 15529 ms  {Loader map objects (synchronizer)}`

**Comment by Chumva** [comment]
I tested favourites (260 points, 50KB), with debug build initialization taking 2-3s, release build 50-80ms. Looks like this issue is related to some other data initialization. I will add more steps to app initialization logs for further investigation

**Comment by sonora** [comment]
Sure, here you go:

[logcat_2024-03-28.log](https://github.com/osmandapp/OsmAnd/files/14795318/logcat_2024-03-28.log)

GPX db seems a good guess, looking at `System.out: Initialized GPX_DB_INITIALIZED in 13568 ms`. :wink:

While GPX has always been slow (see related issues), please note that the observation in this issue is a clear regression: When I revert back to my March 15 build, the startup is immediately fast again. So there' s something we did between 03-15 and 03-22.

**Comment by vshcherb** [comment]
> I'm mainly wondering why all the newer builds have the net.osmand:native: Native File not initialized from cache: occurrences, the March 15 logfile has none of that...

There was a cache added in C++ but that's for iOS cause in Android it was probably not needed however it's tricky may be another issue.

**Comment by vshcherb** [comment]
You should see something like this
04-01 17:33:52.732  5227  5312 I net.osmand:native: Native Cache file initialized: /storage/emulated/0/Android/obb/net.osmand.plus/ind.cache 5

You probably need to remove cache and/or redownload map. Obviously it slows down startup

**Comment by sonora** [comment]
- Deleting _ind.cache_ had no effect - it got re-cteated with the same size (171kB).
- Deleting _ind-core.cache_ (154kB) had no effect either - never got re-created so far.. 
- Redownladed a number of maps - no effect.

- Reverting back to March 15 build - things are fast again.

**Comment by vshcherb** [comment]
@sonora fast again with maps (except GPX?) in the logs. We didn't reproduce after download with new version. You need to delete/redownload specific maps cause I think you have more than 4 specific maps which are broken with native cache

**Comment by sonora** [comment]
@vshcherb Just compiled and tried: Time to see the map screen with map was 27 sec (compares to 7 sec when reverting to my March 15 build). Both times taken on the same device with the same map folder. All my maps are now March 2023 Editions, but that did not fix it. So you say I have to wait for the April editions and try again?

**Comment by Chumva** [comment]
@sonora can you send new logs?

**Comment by sonora** [comment]
Sure, here you go: [logcat_4.7.5_2024-04-02.log.txt](https://github.com/osmandapp/OsmAnd/files/14842384/logcat_4.7.5_2024-04-02.log.txt)

**Comment by vshcherb** [comment]
Something to work with ``` GpxDbHelper Time to loadGpxItems 9566 ms items count 3432 {Initializing app}``` hopefully will get it down with db optimization

**Comment by sonora** [comment]
Yes... it's unclear why the app would load/initialize all my 3432 GPX tracks at every app start?

Isn't that something which should be cached? And as long as I don't use them, couldn't they just sit in the file system, perhaps in the background checked for changes to be updated in the GPX db? And worse: I have a feeling it does that load/init again at every entering of the MyTracks or "Visible on map" Tracks screens?

PS: Number of tracks 3432 is wrong anyway, because this is the **number of files** in the tracks folder regardless of file type, but not all of these are GPX-tracks (some are actually descriptive PDFs and TXT files ...)

**Comment by vshcherb** [comment]
Loading is fine the speed is not fine, 3ms to read 1 entry from DB is too slow. 
We load GPX tracks cause they should appear fast in Configure Map

**Comment by sonora** [comment]
> We load GPX tracks cause they should appear fast in Configure Map

But they don't. Displaying  'Configure maps>Tracks' takes 12-15 sec in my setup (even when no tracks are selected to be visible on map), making me believe that the 'loading' is repeated 'ive' every time that screen is entered.

BTW, this was no different in v4.6/March 15 builds, so whatever we did since then has only worsened the startup perfomance for no benefitI can detect.

**Comment by vshcherb** [comment]
@sonora Maps were related to Cache (https://github.com/osmandapp/OsmAnd-core-legacy/commit/f2470513b191d17f72b90adb4ce4ef38d02e76ae https://github.com/osmandapp/OsmAnd-core-legacy/commit/d885009d2fbd3a3f794b41b1fc85dd7d9419ef92) - 20-22 March

I don't know about 15 March...

**Comment by vshcherb** [comment]
Today going to investigate put more logs let's see

**Comment by Chumva** [comment]
@sonora can you check start time after new changes?

**Comment by sonora** [comment]
Well, it's not really noticeably faster, I have to say... [logcat_2024-04-12.log.txt](https://github.com/osmandapp/OsmAnd/files/14962668/logcat_2024-04-12.log.txt)

The bigger contributions continue to be several occurrences of 
`Native File not initialized from cache:`
(but I still use March maps), and 
```
04-12 21:05:22.353 18706 20078 I net.osmand: GpxDbHelper Time to loadGpxItems 10738 ms items count 3444 {Initializing app}
04-12 21:05:22.400 18706 20078 I net.osmand: GpxDbHelper Time to loadGpxDirItems 46 ms items count 3444 {Initializing app}
04-12 21:05:22.402 18706 20078 I System.out: Initialized GPX_DB_INITIALIZED in 10786 ms
04-12 21:05:22.474 18706 20078 I System.out: Initialized POI_FILTERS_INITIALIZED in 74 ms
```

**Comment by vshcherb** [comment]
First of all we have
```
04-12 21:05:11.614 18706 20078 I System.out: Initialized FAVORITES_INITIALIZED in 714 ms
04-12 21:05:22.402 18706 20078 I System.out: Initialized GPX_DB_INITIALIZED in 10786 ms
```

vs
```
04-02 20:30:53.539 12181 14579 I System.out: Initialized FAVORITES_INITIALIZED in 1376 ms
04-02 20:31:10.293 12181 14579 I System.out: Initialized GPX_DB_INITIALIZED in 16751 ms
```
Original.  So it's speed up

Second I don't see in your logs output for every 100 tracks, which is strange looks like pull request  applied partially...
https://github.com/osmandapp/OsmAnd/pull/19532/files#diff-5d0669429779b6b737660b37ebfb38ce7e01a8cca6a11372a3f85cab28b58fa2R105

**Comment by vshcherb** [comment]
Native initialization goes concurrently and can obviously slow down the process cause it's not necessarily done in parallel on the phone itself. Obviously until native not being initialized, you can't see the map and it takes ~ 20 seconds.
```
04-12 21:05:19.763 18706 20074 W net.osmand:native: Native File not initialized from cache: /storage/3939-6139/osmand/Germany_nordrhein-westfalen_europe.obf 4209 ms
```

I tried to fix possible issue with it https://github.com/osmandapp/OsmAnd-core-legacy/commit/88aaf122df27501630e080d3a0d4d994f407265a

**Comment by vshcherb** [comment]
I just double checked again and the proble is in 
```
04-12 21:05:10.900 18706 20078 I net.osmand:native: Native Cache file initialized: /storage/3939-6139/osmand/ind.cache ```

This file should be updated each time new map is downloaded issue that some maps are not in this file or it's not correctly read. Please delete/ redownload the map  and share logs (probably we could see something there) or share the file itself.

We tried to get these messages but here what I ahve
```
04-13 12:49:11.388 25299 25351 I net.osmand: GpxDbHelper Time to loadGpxItems 354 ms items count 538 {Initializing app}
04-13 12:49:11.394 25299 25299 D SurfaceView@d21eb3f: 220326719 updateSurface: has no frame
04-13 12:49:11.409 25299 25308 I net.osmand.plus: Background concurrent copying GC freed 2130822(93MB) AllocSpace objects, 224(19MB) LOS objects, 39% free, 36MB/60MB, paused 47us,401us total 302.531ms
04-13 12:49:11.413 25299 25299 D SurfaceView@d21eb3f: 220326719 updateSurface: has no frame
04-13 12:49:11.420 25299 25351 I net.osmand: GpxDbHelper Time to loadGpxDirItems 32 ms items count 538 {Initializing app}
04-13 12:49:11.420 25299 25351 I System.out: Initialized GPX_DB_INITIALIZED in 386 ms
04-13 12:49:10.311 25299 25398 I net.osmand: CachedOsmandIndexes Initialize cache ind.cache 19 ms {pool-7-thread-1}
04-13 12:49:10.916 25299 25351 I net.osmand:native: Native Cache file initialized: /storage/emulated/0/Android/obb/net.osmand.plus/ind.cache 2
04-13 12:49:10.917 25299 25351 I System.out: Initialized NATIVE_INITIALIZED in 409 ms

```

**Comment by vshcherb** [comment]
With new logs we will see batches like 
```
04-13 13:11:41.475 29230 29282 I net.osmand: GpxDbHelper Loading tracks batch. took 58ms {Initializing app}
04-13 13:11:41.477 29230 29282 I net.osmand: GpxDbHelper Loading tracks batch. took 2ms {Initializing app}
04-13 13:11:41.479 29230 29282 I net.osmand: GpxDbHelper Loading tracks batch. took 2ms {Initializing app}
```

So please share new log with this parts (@sonora )

**Comment by vshcherb** [comment]
**To do** @Chumva / @Corwin-Kh :
- [ ]  Load new GPX tracks from FS in background only when user enters My places (data will be refreshed only when reenters or enters folders)
- [ ] On entering My Places / Configure Map data should be displayed immediately from Memory (cause track items are loaded during start) - NO DB, NO FS.
- [ ]  Pull to refresh could or could not wait full read from DB or from FS - depends on implementation - please specify how it works

**Comment by sonora** [comment]
Yes, of course I agree it's become measurably faster in recent days! But we have this overall development of startup time, so we are still factors slower than a month ago:
- March 15 build:  **7 sec** _(still reproducible today, I get this consistently once falling back to that build)_
- March 22 build: **40 sec**
- April 14 build: **29 sec**

Here's the new log: [logcat_2024-04-14.log.txt](https://github.com/osmandapp/OsmAnd/files/14969987/logcat_2024-04-14.log.txt). Looks like the first `Loading tracks batch.` occurrence alone
```
04-14 10:15:45.455 17413 20078 I net.osmand: GpxDbHelper Loading tracks batch. took 6878ms {Initializing app}
```
takes **6878ms**, while all others take about 100-150 ms each, totaling **4338 ms**.

**Comment by vshcherb** [comment]
@sonora it's hard to explain what happened for March 15. I suspect only that you used (release or non-debug version) cause in this log I see (start at 10:15:30)
```
04-14 10:15:35.781 17413 20078 I net.osmand:native: JNI_OnLoad completed
04-14 10:15:37.235 17413 20078 I net.osmand:native: Font path /storage/3939-6139/osmand/fonts/05_NotoSans-Regular.ttf index 0
```

So there is no chance to speed up something before even JNI load and indeed for release versions JNI loads much faster.
```
04-14 10:15:32.070 17413 17413 I System.out: Time to start application 1312 ms. Should be less < 800 ms
04-14 10:15:32.148 17413 17413 I System.out: Time to init plugins 78 ms. Should be less < 800 ms
04-14 10:15:32.428 17413 20078 I System.out: Initialized INIT_RENDERERS in 358 ms
04-14 10:15:32.609 17413 20078 I System.out: Initialized POI_TYPES_INITIALIZED in 181 ms
04-14 10:15:32.703 17413 20078 I System.out: Initialized ASSETS_COPIED in 94 ms
04-14 10:15:33.515 17413 20078 I System.out: Initialized MAPS_INITIALIZED in 812 ms
04-14 10:15:33.515 17413 20078 I System.out: Initialized INDEXES_RELOADED in 0 ms
04-14 10:15:34.072 17413 17413 I System.out: Initialized ROUTING_CONFIG_INITIALIZED in 557 ms
04-14 10:15:35.747 17413 20078 I System.out: Initialized TRAVEL_INITIALIZED in 1675 ms
04-14 10:15:37.910 17413 20078 I System.out: Initialized NATIVE_INITIALIZED in 2162 ms
04-14 10:15:38.575 17413 20078 I System.out: Initialized FAVORITES_INITIALIZED in 666 ms
04-14 10:15:49.888 17413 20078 I System.out: Initialized GPX_DB_INITIALIZED in 11312 ms
04-14 10:15:49.961 17413 20078 I System.out: Initialized POI_FILTERS_INITIALIZED in 74 ms
04-14 10:15:51.823 17413 20078 I System.out: Initialized INDEX_REGION_BOUNDARIES in 1862 ms
04-14 10:15:51.911 17413 20078 I System.out: Initialized LOAD_GPX_TRACKS in 88 ms
04-14 10:15:51.916 17413 20078 I System.out: Initialized SAVE_GPX_TRACKS in 5 ms
04-14 10:15:51.993 17413 20078 I System.out: Initialized MARKERS_GROUPS_SYNCED in 77 ms
04-14 10:15:52.226 17413 20078 I System.out: Initialized SEARCH_UI_CORE_INITIALIZED in 233 ms
04-14 10:15:52.226 17413 20078 I System.out: Initialized LIVE_UPDATES_ALERTS_CHECKED in 0 ms
04-14 10:15:53.035 17413 20078 I System.out: Initialized BROUTER_INITIALIZED in 809 ms
```

**Comment by vshcherb** [comment]
Checked files as well have no idea: why some files are not initialized there is no practically difference with others... I added more logs with core-legacy commit to see what's the difference may be in file size.
```
04-14 10:15:46.842 17413 20074 I net.osmand:native: Native open file: /storage/3939-6139/osmand/Germany_brandenburg_europe.obf
04-14 10:15:48.057 17413 20074 W net.osmand:native: Native File not initialized from cache: /storage/3939-6139/osmand/Germany_brandenburg_europe.obf 1216 ms
04-14 10:15:48.059 17413 20074 I net.osmand:native: Native open file: /storage/3939-6139/osmand/Germany_bayern_europe.obf
04-14 10:15:53.825 17413 20074 W net.osmand:native: Native File not initialized from cache: /storage/3939-6139/osmand/Germany_bayern_europe.obf 5768 ms
```

```
Germany_thueringen_europe.obf Fri Mar 01 02:00:00 EET 2024 314991633
Germany_sachsen_europe.obf Fri Mar 01 02:00:00 EET 2024 520116620
Germany_sachsen-anhalt_europe.obf Fri Mar 01 02:00:00 EET 2024 306043179
Germany_rheinland-pfalz_europe.obf Fri Mar 01 02:00:00 EET 2024 530610982
Germany_nordrhein-westfalen_europe.obf Fri Mar 01 02:00:00 EET 2024 1914331266
Germany_hessen_europe.obf Fri Mar 01 02:00:00 EET 2024 745161582
Germany_europe.srtm.obf Mon Nov 15 11:33:32 EET 2021 275600406
Germany_brandenburg_europe.obf Fri Mar 01 02:00:00 EET 2024 585757445
Germany_bayern_europe.obf Fri Mar 01 02:00:00 EET 2024 1685935184
Germany_bayern_europe.wiki.obf Thu Dec 07 13:08:17 EET 2023 391016192
Germany_baden-wuerttemberg_europe.obf Fri Mar 01 02:00:00 EET 2024 1370636075
```

**Comment by sonora** [comment]
Here's the new log: [logcat_2024-04-15.log.txt](https://github.com/osmandapp/OsmAnd/files/14982016/logcat_2024-04-15.log.txt). I am also running out of ideas...

I notice that above the `JNI_OnLoad completed` line we always have 
```
linker  : Warning: "/data/app/~~9FBAqP9FNlf-uNm7d81fUw==/net.osmand.plus-gt1v8_fJYgqK4NPRWW9pgg==/base.apk!/lib/arm64-v8a/libc++_shared.so" unused DT entry: unknown processor-specific (type 0x70000001 arg 0x0) (ignoring)
```
which looks a bit strange but may not be an issue here.

Favorites and partially GPX init seem improved since March 15, but apparently **none of the 10 Native Files the apps initializes at startup seems to be initialized from cache?** So I really guess one of the core-legacy commits on March 20-22 about caching must cause this.

I am strictly using osmandapp/master code with the one exception that I grant the app "All files access" `("android.permission.MANAGE_EXTERNAL_STORAGE")` to use sd-card/osmand as data folder. But I can't see how the commits of March 20-22 have affected what has worked flawlessly for years before...

**Comment by sonora** [comment]
One more comment: It's not just the app startup which has slowed after March 15 builds: It seems to happen during zooming or panning every time a new obf file is needed, e.g. when I zoom out to World Overview level, I get 30 sec of not updating the map screen. Or when I zoom in to a country on another continent (where I have the obf file), it app also stalls for a long time (and the log shows the according "Native File not initialized from cache:" message.)

(Should caching occur at all in Android? I believe the commits of March 20-22 were perhaps targeted at iOS?)

**Comment by vshcherb** [comment]
Yes targeted only iOS and I didn't see how they are correlated to be honest, Cache on Android still is created by Java and it is read by Native as before

**Comment by sonora** [comment]
So, in summary:
- We are still in the dark why builds between March 15 and March 22 have introduced such a significant performance issue at all (100% reproducible when switching between builds)? 
- We do not know what causes the ominous `net.osmand:native: Native File not initialized from cache:` log entries.
- Do we even know why a logcat does show any "cache" related lines only when using the V1 engine, not for V2? Is this expected, or which of the two cases is the "anomaly"?

-  [ ] PS: I still believe that https://github.com/osmandapp/OsmAnd/issues/19373#issuecomment-2053602225 is an optimization worth implementing, even if is not the primary factor behind this issue. (BTW: The Favorites init we seem to have fixed well!)

**Comment by sonora** [comment]
Upon closer investigation, this fixes it and produces "fast" builds for me again:

![image](https://github.com/osmandapp/OsmAnd/assets/845735/9ce3aec4-5044-4a51-a7ee-5a4f962212d1)

So I guess there is an issue with versioning the cache, `c->version()` probably does not deliver what is expected? Am I wrong, or does `c` not have a version() at this point, as it is always only created 3 lines above in `OsmAnd::OBF::OsmAndStoredIndex* c = new OsmAnd::OBF::OsmAndStoredIndex();` ?

**Comment by sonora** [comment]
Haha, that fixes it: https://github.com/osmandapp/OsmAnd-core-legacy/commit/2093630a57267f4bd67843e1e7f703fe322ccd91:

But it is probably not a good idea to hard-code a CACHE_VERSION number here, while the actual VERSION written to storage is hardcoded separately in the OsmAndStoredIndex class? Or else thing will break again with the next VERSION upgrade?

-  [ ] So other than that, remaining open here as an optimization still worth implementing is: https://github.com/osmandapp/OsmAnd/issues/19373#issuecomment-2053602225
-  [x] The Favorites init we seem to have fixed well!)

**Comment by vshcherb** [comment]
Yes,
- [ ]  Refactoring is needed anyway to have good undestanding what's going on (https://github.com/osmandapp/OsmAnd/issues/19373#issuecomment-2053602225)
- [ ] Good finding ! We need to check CACHE_VERSION for IOS and probably ignore for Android or combine correclty

**Comment by ivanPyrohivskyi** [comment]
Merged CACHE_VERSION = 3 to [r4.7](https://github.com/osmandapp/OsmAnd-core-legacy/commits/r4.7/)
Checked on iOS. Cache file was rewritten successfully (from version 1 to version 3).
Checked on Android, works fine: 
```Native Cache file initialized: /storage/emulated/0/Android/data/net.osmand.plus/files/ind.cache 0```
```Native file initialized from cache: /storage/emulated/0/Android/data/net.osmand.plus/files/Ukraine_europe.obf 0 ms```

**Comment by sonora** [comment]
Seems to me like the recent expansion and bump of the GPX database to version 21 has further aggravated the GPX portion of this issue.

 I experience the fastest response of entering the MyTracks screen (<2sec) in the initial phase after a fresh install, potentially before the GPX database has been built, could it be?

**Comment by vshcherb** [comment]
Merged to https://github.com/osmandapp/OsmAnd/issues/19683

**Comment by vshcherb** [comment]
This issue will be closed as too long and fixed. the rest refactorlng will be separated

**Comment by sonora** [comment]
We marked this as fixed, but shall we not make the cache version a global parameter in the app instead of maintaining it in separate methods, see e.g. also here https://github.com/osmandapp/OsmAnd/commit/5f852a02f03f1b13b2bbcc53baeab450f77e044e?

**Comment by ivanPyrohivskyi** [comment]
Added comments to code for cache version number synchronisation (java and core-legacy).
https://github.com/osmandapp/OsmAnd/commit/f1aa507175e4f549469c7734f46c3e5e093a70e2
https://github.com/osmandapp/OsmAnd-core-legacy/commit/8f87567825c6989ddb86537c5e56bdeb1e52f2f7

Java and legacy has different approach - OOP and functional, so better to left as is.
